CacheManager.StackExchange.Redis.SelfCleaning.Tests/SelfCleaningRedisIntegrationTests.cs

```diff
@@ -0,0 +1,247 @@
+                .Returns<byte[], RedisKey[], RedisValue[], CommandFlags>((hash, keys, values, flags) =>
+                    RedisResult.Create(_fauxDatabase[keys[0]].Values));
+
+            // Remove
+            _databaseMock
+                .Setup(database => database.KeyDelete(It.IsAny<RedisKey>(), It.IsAny<CommandFlags>()))
+                .Callback<RedisKey, CommandFlags>((key, flags) => _fauxDatabase.Remove(key));
+        }
+
+        private void SetupConnectionMock()
+        {
+            _connectionMock.SetupGet(connection => connection.Configuration).Returns("connectionString");
+            _connectionMock.Setup(connection => connection.GetEndPoints(It.IsAny<bool>())).Returns(new EndPoint[1]);
+            _connectionMock.Setup(connection => connection.GetServer(It.IsAny<EndPoint>(), It.IsAny<object>()))
+                .Returns(_serverMock.Object);
+            _connectionMock.Setup(connection => connection.GetDatabase(It.IsAny<int>(), It.IsAny<object>()))
+                .Returns(_databaseMock.Object);
+        }
+
+        private void SetupCache()
+        {
+            double cleanupInterval = 100;
+            var cleanupTimer = new DefaultTimer(cleanupInterval);
+            _timeToLive = TimeSpan.FromSeconds(1);
+
+            _cache = CacheFactory.Build<DummyModel>(part => part
+                .WithProtoBufSerializer()
+                .WithSelfCleaningRedisConfiguration(_connectionMock.Object, cleanupTimer, _timeToLive,
+                    out string configurationKey)
+                .WithSelfCleaningRedisCacheHandle(configurationKey));
+
+            _onRemoveByHandleInvocations = new List<(CacheItemRemovedEventArgs, DateTime)>();
+            _cache.OnRemoveByHandle += (sender, args) => _onRemoveByHandleInvocations.Add((args, DateTime.Now));
+
+            _cache.CacheHandles.OfType<IStartable>().Single().Start();
+        }
+
+        #endregion
+
+        [TearDown]
+        public void TearDown()
+        {
+            _fauxDatabase.Clear();
+            _onRemoveByHandleInvocations.Clear();
+        }
+
+        [OneTimeTearDown]
+        public void OneTimeTearDown()
+        {
+            _cache.Dispose();
+        }
+
+        [TestCase(1, false)]
+        [TestCase(2, true)]
+        [TestCase(3, false)]
+        [TestCase(4, true)]
+        [TestCase(5, false)]
+        public void Put_AddSomeAmountOfKeys_OnRemoveByHandleCalledAfterTimeToLiveWithCorrectValueAndReason(
+            int amountOfKeys, bool delayBetweenInsertions)
+        {
+            // Arrange
+            IEnumerable<(string, DummyModel)> keys = Enumerable.Repeat("key", amountOfKeys)
+                .Select((key, i) => (key + i, new DummyModel {Property = "property" + i}));
+
+            TimeSpan delayBetweenInsertionsSpan = delayBetweenInsertions ? _timeToLive / 2 : TimeSpan.Zero;
+
+            // Act
+            var keysWithInsertionTimes = new List<(string, DummyModel, DateTime)>();
```

Comment:
1. add names for the fields (although they arent being used), it will help understand their meaning. expectedKey, expectedValue, insertionTime.
2. thats key and values and insertionTime... and it should have "expected" in its name.
maybe expectedCachedItems, or expectedCachedItemsWithInsertionTime?

Reply:
changed